File: sesion2-lab/entra-aspnet/entra-demo/Controllers/ApiController.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace EntraDemo.Controllers;

[ApiController]
[Route("api")]
public class ApiController : ControllerBase
{
    [HttpGet("public/hello")]
    [AllowAnonymous]
    public IActionResult PublicHello()
    {
        return Ok(new
        {
            message = "Hola! Este endpoint es publico, no requiere token.",
            scope = "public",
            idp = "Microsoft Entra ID"
        });
    }

    [HttpGet("me")]
    [Authorize]
    public IActionResult Me()
    {
        var username = FirstNonEmpty(
            User.FindFirstValue("preferred_username"),
            User.FindFirstValue("upn"),
            User.FindFirstValue(ClaimTypes.Name),
            User.FindFirstValue("name"),
            User.FindFirstValue(ClaimTypes.NameIdentifier));

        var email = User.FindFirstValue(ClaimTypes.Email)
            ?? User.FindFirstValue("email");

        var roles = User.FindAll("roles").Select(c => c.Value).ToList();
        var authorities = roles.Select(r => "ROLE_" + r).ToList();

        return Ok(new
        {
            message = "Estas autenticado.",
            user = username,
            email,
            authorities,
            roles
        });
    }

    [HttpGet("admin/hello")]
    [Authorize(Roles = "ADMIN")]
    public IActionResult AdminHello()
    {
        var username = FirstNonEmpty(
            User.FindFirstValue("preferred_username"),
            User.FindFirstValue("upn"),
            User.FindFirstValue(ClaimTypes.Name),
            User.FindFirstValue("name"),
            User.FindFirstValue(ClaimTypes.NameIdentifier));

        return Ok(new
        {
            message = "Bienvenido al area de administracion.",
            user = username,
            scope = "admin"
        });
    }

    private static string? FirstNonEmpty(params string?[] values)
    {
        foreach (var value in values)
        {
            if (!string.IsNullOrWhiteSpace(value))
            {
                return value;
            }
        }
        return null;
    }
}

File: sesion2-lab/pkce-client-keycloak/resource-server/src/main/java/com/example/demo/web/ApiController.java
package com.example.demo.web;

import java.util.List;
import java.util.Map;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
public class ApiController {

    @GetMapping("/public/hello")
    public Map<String, Object> publicHello() {
        return Map.of(
            "message", "Hola! Endpoint publico del Resource Server.",
            "scope", "public",
            "idp", "Keycloak (realm demo)");
    }

    @GetMapping("/me")
    public Map<String, Object> me(@AuthenticationPrincipal Jwt jwt, Authentication authentication) {
        List<String> authorities = authentication.getAuthorities().stream()
            .map(GrantedAuthority::getAuthority)
            .toList();
        return Map.of(
            "message", "Estas autenticado con un JWT obtenido via PKCE.",
            "user", jwt.getClaimAsString("preferred_username"),
            "email", String.valueOf(jwt.getClaimAsString("email")),
            "authorities", authorities);
    }

    @GetMapping("/admin/hello")
    public Map<String, Object> adminHello(@AuthenticationPrincipal Jwt jwt) {
        return Map.of(
            "message", "Bienvenido al area de administracion.",
            "user", jwt.getClaimAsString("preferred_username"),
            "scope", "admin");
    }
}
